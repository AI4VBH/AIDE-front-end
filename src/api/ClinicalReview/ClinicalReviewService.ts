import { createAxiosInstance } from "@/utils/axios-helpers";

const http = createAxiosInstance();

export async function getStudy(executionId: string) {
    const response = await http.get(`/clinical-review/${executionId}`);

    return response.data;
}
