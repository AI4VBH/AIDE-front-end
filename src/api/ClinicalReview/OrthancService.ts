/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { createAxiosOrthancInstance } from "@/utils/axios-helpers";

const httpService = createAxiosOrthancInstance();

export async function findStudy(StudyInstanceUID: string): Promise<any> {
    const response = await httpService.post(`/tools/find`, {
        Level: "Study",
        Expand: true,
        Limit: 101,
        Query: {
            StudyDate: "",
            StudyInstanceUID: StudyInstanceUID,
        },
        Full: true,
    });
    return response.data;
}

export async function getStudy(StudyId: string): Promise<any> {
    const response = await httpService.get(`/studies/${StudyId}`);
    return response.data;
}

export async function getSeries(StudyId: string): Promise<any> {
    const response = await httpService.get(`/series/${StudyId}`);
    return response.data;
}

export async function getSeriesOrderedSlices(StudyId: string): Promise<any> {
    const response = await httpService.get(`/series/${StudyId}/ordered-slices`);
    return response.data;
}

export async function getInstance(StudyId: string): Promise<any> {
    const response = await httpService.get(`/instances/${StudyId}`);
    return response.data;
}

export async function getInstanceMetadata(StudyId: string): Promise<any> {
    const response = await httpService.get(`/instances/${StudyId}/tags`);
    return response.data;
}

export async function downloadStudy(StudyId: string): Promise<any> {
    const response = await httpService.get(`/studies/${StudyId}/media`, { responseType: "blob" });
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement("a");
    link.href = url;
    link.setAttribute("download", `${StudyId}.zip`); //or any other extension
    document.body.appendChild(link);
    link.click();
    return response.data;
}
