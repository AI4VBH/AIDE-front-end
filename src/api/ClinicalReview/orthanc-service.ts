import { createAxiosOrthancInstance } from "@/utils/axios-helpers";

const httpService = createAxiosOrthancInstance();

export interface ExpandedDicomTags {
    [key: string]: { Name: string; Type: string; Value: string };
}

export interface SimpleDicomTags {
    [key: string]: string;
}

interface FindResultItem {
    ID: string;
    IsStable: boolean;
    LastUpdate: string;
    ParentPatient: string;
    Series: string[];
    Type: string;
}

export interface SimpleFindResultItem extends FindResultItem {
    MainDicomTags: SimpleDicomTags;
    PatientMainDicomTags: SimpleDicomTags;
}

export async function findStudyMetadata(StudyInstanceUID: string): Promise<SimpleFindResultItem[]> {
    const response = await httpService.post<SimpleFindResultItem[]>(`/tools/find`, {
        Level: "Study",
        Expand: true,
        Limit: 101,
        Query: {
            StudyDate: "",
            StudyInstanceUID: StudyInstanceUID,
        },
        Full: false,
    });

    return response.data;
}
