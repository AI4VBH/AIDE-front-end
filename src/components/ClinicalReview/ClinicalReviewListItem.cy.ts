import ClinicalReviewListItem from "./ClinicalReviewListItem.vue";

describe("<ClinicalReviewListItem />", () => {
    it("renders clinical review item with clinical review task data", () => {
        // see: https://test-utils.vuejs.org/guide/
        const task = {
            _id: "678",
            clinical_review_message: {
                task_id: "",
                reviewed_task_id: "cde",
                execution_id: "677",
                reviewed_execution_id: "abc",
                correlation_id: "123",
                workflow_name: "bobwf",
                patient_metadata: {
                    patient_name: "Joe Batt",
                    patient_id: "1299-123-232-3422",
                    patient_sex: "M",
                    patient_Dob: "10-10-2000",
                    patient_Age: "23",
                },
                files: [],
                reviewer_roles: ["admin", "clinician"],
                application_metadata: {
                    application_name: "stroke model",
                    application_version: "1.1",
                    application_mode: "CU",
                },
            },
            reviewed: false,
            received: new Date(),
        };
        cy.mount(ClinicalReviewListItem as any, {
            propsData: {
                patient_name: task.clinical_review_message.patient_metadata.patient_name,
                patient_id: task.clinical_review_message.patient_metadata.patient_id,
                patient_age: task.clinical_review_message.patient_metadata.patient_Age,
                patient_sex: task.clinical_review_message.patient_metadata.patient_sex,
                application_name:
                    task.clinical_review_message.application_metadata.application_name,
                application_version:
                    task.clinical_review_message.application_metadata.application_version,
                application_mode:
                    task.clinical_review_message.application_metadata.application_mode,
                received: task.received,
            },
        });

        cy.get("[data-cy=patient-name]").should(
            "contain.text",
            task.clinical_review_message.patient_metadata.patient_name,
        );
        cy.get("[data-cy=patient-id]").should(
            "contain.text",
            task.clinical_review_message.patient_metadata.patient_id,
        );
        cy.get("[data-cy=patient-age]").should(
            "contain.text",
            task.clinical_review_message.patient_metadata.patient_Age,
        );
        cy.get("[data-cy=patient-sex]").should(
            "contain.text",
            task.clinical_review_message.patient_metadata.patient_sex,
        );
        cy.get("[data-cy=application]").should(
            "contain.text",
            task.clinical_review_message.application_metadata.application_name,
            task.clinical_review_message.application_metadata.application_version,
        );
        cy.get("[data-cy=mode]").should(
            "contain.text",
            task.clinical_review_message.application_metadata.application_mode,
        );
        cy.get("[data-cy=received]").should("contain.text", task.received);
    });
});
