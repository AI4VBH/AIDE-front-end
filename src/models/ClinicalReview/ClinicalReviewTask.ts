export interface ClinicalReviewSeries {
    series_id: string;
    modality: string;
    files: string[];
}

export interface ClinicalReviewTaskDetail {
    study: ClinicalReviewSeries[];
}
