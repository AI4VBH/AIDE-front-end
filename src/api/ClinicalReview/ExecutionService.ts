/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Execution, ExecutionPage } from "@/models/ClinicalReview/Execution";
import { ExecutionStat } from "@/models/ClinicalReview/ExecutionStat";
import { createAxiosInstance, ErrorMessageMap } from "@/utils/axios-helpers";

const executionsErrorMessages: ErrorMessageMap = {
    get: "Something unexpected went wrong retrieving the executions",
};

const http = createAxiosInstance(executionsErrorMessages);

export async function getAllExecutionsPage(
    from: string,
    to: string,
    approved: string,
): Promise<ExecutionPage> {
    const response = await http.get(`/executions?from=${from}&to=${to}&approved=${approved}`);
    return response.data;
}

export async function getAllModelExecutions(
    model_id: string,
    from: string,
    to: string,
): Promise<ExecutionPage> {
    const response = await http.get(`/executions?model_id=${model_id}&from=${from}&to=${to}`);
    return response.data;
}

export async function getExecutionPipelines(correlation_id: string): Promise<Array<Execution>> {
    const response = await http.get(`/pipeline/${correlation_id}`);
    return response.data;
}

export async function getExecutionStats(days: string): Promise<ExecutionStat> {
    const response = await http.get(`/execution_stats?days=${days}`);
    return response.data;
}

export async function getModelExecutionStats(
    days: string,
    model_id: string,
): Promise<ExecutionStat> {
    const response = await http.get(`/execution_stats?days=${days}&model_id=${model_id}`);
    return response.data;
}

export async function getFile(file_path: string): Promise<any> {
    const response = await http.post("/file", { file_path: file_path }, { responseType: "blob" });
    if (response.headers["content-type"] === "application/zip") {
        file_path = file_path + ".zip";
    }
    const file_name = file_path.split("/");
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement("a");
    link.href = url;
    link.setAttribute("download", file_name[file_name.length - 1]); //or any other extension
    document.body.appendChild(link);
    link.click();
    return response.data;
}
