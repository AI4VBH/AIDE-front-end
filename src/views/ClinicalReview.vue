<!--
  Copyright 2022 Crown Copyright

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
  -->

<template>
    <v-container fluid class="clinical-review-container">
        <patient-header
            v-if="taskCount > 0 && currentTaskClinicalReviewMessage"
            :patient-metadata="currentTaskClinicalReviewMessage.patient_metadata"
            :study-date="studyDate"
            @task-accepted="acceptTask"
            @task-rejected="rejectTask"
        />
        <v-row class="clinical-review">
            <v-col class="task-list" v-show="!tasksLoading && taskCount > 0">
                <clinical-review-task-list
                    @task-selected="taskSelected"
                    @tasks-count-updated="taskCountUpdated"
                    @tasks-loading-changed="tasksLoadingChanged"
                >
                    <template v-slot="{ throttledFetchTasks }">
                        <v-dialog :value="actionModal" persistent max-width="500px">
                            <accept-reject-dialog
                                v-if="currentTaskClinicalReviewMessage"
                                :open="actionModal"
                                :reject="reject"
                                :application-name="
                                    currentTaskClinicalReviewMessage.application_metadata
                                        .application_name
                                "
                                @cancel="actionModal = false"
                                @perform-action="(data) => performAction(data, throttledFetchTasks)"
                            />
                        </v-dialog>
                    </template>
                </clinical-review-task-list>
            </v-col>
            <v-col class="dicom-view">
                <dicom-view
                    v-if="currentTaskExecutionId"
                    :task-execution-id="currentTaskExecutionId"
                    @study-selected="studySelected"
                />

                <div
                    v-if="!currentTaskExecutionId && !tasksLoading && taskCount === 0"
                    class="no-tasks-left"
                >
                    <v-icon x-large color="success">mdi-checkbox-marked-circle-outline</v-icon>
                    <p class="mt-2">There are no application outputs left to review</p>
                </div>
            </v-col>
        </v-row>
    </v-container>
</template>

<script lang="ts">
import Vue from "vue";
import { defineComponent } from "vue";
import DicomView from "@/components/clinical-review/dicom-view.vue";
import ClinicalReviewTaskList from "@/components/clinical-review/tasks/task-list.vue";
import { updateClinicalReview } from "../api/ClinicalReview/ClinicalReviewService";
import PatientHeader from "@/components/clinical-review/patient-header.vue";
import AcceptRejectDialog from "@/components/clinical-review/accept-reject-dialog.vue";
import {
    ClinicalReviewRecord,
    ClinicalReviewStudyDetails,
    ClinicalReviewTaskDetails,
} from "@/models/ClinicalReview/ClinicalReviewTask";

type ClinicalReviewData = {
    currentTaskExecutionId?: string;
    currentTaskClinicalReviewMessage?: ClinicalReviewTaskDetails;
    taskCount: number;
    tasksLoading: boolean;
    actionModal: boolean;
    reject: boolean;
    studyDate?: string;
};

export default defineComponent({
    metaInfo: {
        title: "Clinical Review",
    },
    components: {
        DicomView,
        ClinicalReviewTaskList,
        PatientHeader,
        AcceptRejectDialog,
    },
    data(): ClinicalReviewData {
        return {
            currentTaskExecutionId: "",
            currentTaskClinicalReviewMessage: undefined,
            studyDate: "",
            taskCount: 0,
            tasksLoading: true,
            actionModal: false,
            reject: false,
        };
    },
    methods: {
        taskSelected(execution_id: string, task: ClinicalReviewRecord) {
            this.currentTaskExecutionId = execution_id;
            this.currentTaskClinicalReviewMessage = task.clinical_review_message;
        },
        taskCountUpdated(count: number) {
            this.taskCount = count;
        },
        studySelected(study: ClinicalReviewStudyDetails) {
            this.studyDate = study.study_date;
        },
        tasksLoadingChanged(loading: boolean) {
            this.tasksLoading = loading;
        },
        acceptTask() {
            this.actionModal = true;
            this.reject = false;
        },
        rejectTask() {
            this.actionModal = true;
            this.reject = true;
        },
        async performAction(
            data: { reason: string | undefined; description: string; acceptance: boolean },
            fetchTasks: () => void,
        ) {
            if (!this.currentTaskExecutionId) {
                return;
            }

            const responseOk = await updateClinicalReview(
                this.currentTaskExecutionId,
                data.acceptance,
                data.description,
                data.acceptance ? undefined : data.reason,
            );

            if (responseOk) {
                Vue.$toast.success(
                    `Clinical Review has been ${!data.acceptance ? "rejected" : "accepted"}`,
                );
                this.actionModal = false;
                fetchTasks();
            }
        },
    },
});
</script>

<style lang="scss" scoped>
.headerLabel {
    color: #424242;
    font-weight: bold;
    line-height: 24px;
    vertical-align: middle;
}

.clinical-review-container {
    height: 100%;
    padding: 0;
    overflow: hidden;
    display: flex;
    flex-direction: column;

    > .row {
        margin: 0;
    }
}

.patient-header {
    background-color: #fbfbfb;
    height: 80px;
}

.clinical-review {
    margin: 0;
    height: 100%;
    min-width: 1000px;

    > .col {
        padding: 0;
    }

    .task-list {
        width: 350px;
        flex: initial;
        display: flex;
    }
}

.no-tasks-left {
    display: flex;
    flex-direction: column;
    height: 100%;
    align-items: center;
    justify-content: center;
}
</style>
