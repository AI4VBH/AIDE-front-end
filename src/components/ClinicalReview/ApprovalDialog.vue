<template>
    <v-dialog transition="dialog-bottom-transition" max-width="60vw" v-model="dialog4">
        <template v-slot:default="dialog4">
            <v-card>
                <v-card-title> {{ title }} </v-card-title>
                <div style="margin: 0px 30px 0px 30px">
                    <v-alert dense type="info"> This action will be signed by you </v-alert>
                    <v-select
                        v-if="!approval"
                        v-model="reason"
                        :items="items"
                        filled
                        label="Reason for rejection"
                        dense
                    ></v-select>
                    <v-text-field
                        v-model="description"
                        label="Description(optional)"
                        filled
                    ></v-text-field>

                    <v-checkbox
                        v-model="checkbox"
                        color="#61366E"
                        :label="'I accept that this is signed under my name'"
                    ></v-checkbox>
                    <v-card-actions class="justify-end">
                        <v-col cols="4" style="color: #fff">
                            <v-btn
                                depressed
                                style="margin-right: 8px"
                                @click="dialog4.value = false"
                                >Cancel</v-btn
                            >
                            <v-btn
                                v-if="approval"
                                data-test="accept-btn"
                                color="#4CAF50"
                                dark
                                :loading="loading"
                                @click="acceptReview()"
                            >
                                Accept
                            </v-btn>

                            <v-btn
                                v-else
                                data-test="reject-btn"
                                color="#DC2626"
                                dark
                                :loading="loading"
                                @click="rejectReview()"
                            >
                                Reject
                            </v-btn>
                        </v-col>
                    </v-card-actions>
                </div>
            </v-card>
        </template>
    </v-dialog>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";
import { EventBus } from "@/event-bus";
import VueJsonPretty from "vue-json-pretty";
import "vue-json-pretty/lib/styles.css";
import { updateClinicalReview } from "../../api/ExecutionService";

@Component({
    components: { VueJsonPretty },
})
export default class ApprovalDialog extends Vue {
    dialog4 = false;
    executionId = "";
    approval = false;
    model_name = "";
    title = "";
    items = ["Other", "Wrong Diagnosis"];
    checkbox = false;
    reason = "";
    description = "";
    loading = false;

    created(): void {
        EventBus.$on(
            "openApprovalDialog",
            async (dialog4: boolean, execution_uid: string, approval: boolean, title: string) => {
                this.dialog4 = dialog4;
                this.executionId = execution_uid;
                this.approval = approval;
                this.title = title;
            },
        );
    }

    async acceptReview() {
        if (this.checkbox) {
            this.loading = true;
            var response = await updateClinicalReview(
                this.executionId,
                "true",
                "",
                this.description,
            );

            if (response.status === true) {
                EventBus.$emit("updateTaskList", this.executionId);
            }
            this.loading = false;
            this.dialog4 = false;
        } else {
            alert("Required fields not filled!");
        }
    }

    async rejectReview() {
        if (this.checkbox && this.reason !== "") {
            this.loading = true;
            var response = await updateClinicalReview(
                this.executionId,
                "false",
                this.reason,
                this.description,
            );
            if (response.status === true) {
                EventBus.$emit("updateTaskList", this.executionId);
            }
            this.loading = false;
            this.dialog4 = false;
        } else {
            alert("Required fields not filled!");
        }
    }
}
</script>
