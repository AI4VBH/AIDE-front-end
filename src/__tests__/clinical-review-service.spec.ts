/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { rest } from "msw";
import { setupServer } from "msw/node";
import { createMock } from "@golevelup/ts-jest";
import { expect, test as it } from "@jest/globals";
import Vue from "vue";
import { ToastApi } from "vue-toast-notification";

import {
    getClinicalReviewTasks,
    getDicomFile,
    getStudy,
    updateClinicalReview,
} from "@/api/ClinicalReview/ClinicalReviewService";
import { getDefaultPagedResponse } from "@/models/common/IPagedResponse";
import { ClinicalReviewRecord } from "@/models/ClinicalReview/ClinicalReviewTask";

jest.mock("vue");

const server = setupServer();
const mockToaster = createMock<ToastApi>();
Vue.$toast = mockToaster;

describe("getStudy", () => {
    beforeAll(() => {
        server.listen();
    });

    afterAll(() => {
        server.close();
    });

    beforeEach(() => {
        server.resetHandlers();
        jest.resetAllMocks();
    });

    describe("getStudy", () => {
        it("returns expected OK result", async () => {
            server.use(
                rest.get(
                    "*/clinical-review/fd04d767-ae5c-441a-87dc-89a91abaa3ec",
                    (_, res, ctx) => {
                        return res(
                            ctx.json({
                                study_date: "2022-11-11T11:11:11",
                                study: [
                                    {
                                        series_uid: "e4350707-2981-4ec2-94a0-caf2a7ce63b9",
                                        modality: "CT",
                                        files: ["FILE0001.dcm"],
                                    },
                                ],
                            }),
                        );
                    },
                ),
            );

            const result = await getStudy("fd04d767-ae5c-441a-87dc-89a91abaa3ec");

            expect(result).toMatchSnapshot();
        });

        it.each([400, 404, 500])(
            "calls toast with the right message when request fails",
            async (status) => {
                server.use(
                    rest.get(
                        "*/clinical-review/bde085c7-e82b-4cf0-9b08-00a23fbb66a0",
                        (_req, res, ctx) => {
                            return res(ctx.status(status));
                        },
                    ),
                );

                try {
                    await getStudy("bde085c7-e82b-4cf0-9b08-00a23fbb66a0");
                } catch {
                    expect(mockToaster.error).toHaveBeenCalledWith(
                        "Something unexpected went wrong retrieving clinical review tasks",
                    );
                }
            },
        );
    });

    describe("getDicomFile", () => {
        it.each([400, 404, 500])(
            "calls toast with the right message when request fails",
            async (status) => {
                server.use(
                    rest.get("*/clinical-review/dicom", (_req, res, ctx) => {
                        return res(ctx.status(status));
                    }),
                );

                try {
                    await getDicomFile("FILE00001.dcm");
                } catch {
                    expect(mockToaster.error).toHaveBeenCalledWith(
                        "Something unexpected went wrong retrieving clinical review tasks",
                    );
                }
            },
        );
    });

    describe("getClinicalReviewTasks", () => {
        it.each([
            [
                1,
                10,
                "patient-id",
                undefined,
                undefined,
                "?pageNumber=1&pageSize=10&patientId=patient-id&patientName=&applicationName=",
            ],
            [
                2,
                5,
                undefined,
                "patient name",
                undefined,
                "?pageNumber=2&pageSize=5&patientId=&patientName=patient+name&applicationName=",
            ],
            [
                2,
                6,
                undefined,
                undefined,
                "app name",
                "?pageNumber=2&pageSize=6&patientId=&patientName=&applicationName=app+name",
            ],
        ])(
            "passes the query params",
            async (pageNumber, pageSize, patientId, patientName, applicationName, query) => {
                server.use(
                    rest.get("*/clinical-review", (req, res, ctx) => {
                        const search = req.url.search;

                        expect(search).toBe(query);

                        const defaultResponse = getDefaultPagedResponse<ClinicalReviewRecord>();

                        return res(ctx.json(defaultResponse));
                    }),
                );

                await getClinicalReviewTasks({
                    pageNumber,
                    pageSize,
                    patientId,
                    patientName,
                    applicationName,
                });
            },
        );

        it.each([400, 404, 500])(
            "calls toast with the right message when request fails",
            async (status) => {
                server.use(
                    rest.get("*/clinical-review", (_req, res, ctx) => {
                        return res(ctx.status(status));
                    }),
                );

                try {
                    await getClinicalReviewTasks({ pageNumber: 1, pageSize: 10 });
                } catch {
                    expect(mockToaster.error).toHaveBeenCalledWith(
                        "Something unexpected went wrong retrieving clinical review tasks",
                    );
                }
            },
        );
    });

    describe("updateClinicalReview", () => {
        it.each([
            [true, "All ok", undefined],
            [false, "Other", "Invalid"],
        ])("returns expected OK result", async (acceptance, description, reason) => {
            server.use(
                rest.put(
                    "*/clinical-review/f1558699-4e19-435d-9789-b86ff6ddd288",
                    async (req, res, ctx) => {
                        const body = await req.json();

                        expect(body).toMatchSnapshot();
                        return res(ctx.status(200));
                    },
                ),
            );

            await updateClinicalReview(
                "f1558699-4e19-435d-9789-b86ff6ddd288",
                acceptance,
                description,
                reason,
            );
        });

        it.each([400, 404, 500])(
            "calls toast with the right message when request fails",
            async (status) => {
                server.use(
                    rest.put(
                        "*/clinical-review/f1558699-4e19-435d-9789-b86ff6ddd288",
                        (_req, res, ctx) => {
                            return res(ctx.status(status));
                        },
                    ),
                );

                try {
                    await updateClinicalReview(
                        "f1558699-4e19-435d-9789-b86ff6ddd288",
                        true,
                        "",
                        "description",
                    );
                } catch {
                    expect(mockToaster.error).toHaveBeenCalledWith(
                        "Something unexpected went wrong saving your review",
                    );
                }
            },
        );
    });
});
