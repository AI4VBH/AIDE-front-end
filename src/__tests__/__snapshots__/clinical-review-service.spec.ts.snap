// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStudy getStudy returns expected OK result 1`] = `
Object {
  "study": Array [
    Object {
      "files": Array [
        "FILE0001.dcm",
      ],
      "modality": "CT",
      "series_uid": "e4350707-2981-4ec2-94a0-caf2a7ce63b9",
    },
  ],
  "study_date": "2022-11-11T11:11:11",
}
`;

exports[`getStudy updateClinicalReview returns expected OK result 1`] = `
Object {
  "acceptance": true,
  "message": "All ok",
}
`;

exports[`getStudy updateClinicalReview returns expected OK result 2`] = `
Object {
  "acceptance": false,
  "message": "Other",
  "reason": "Invalid",
}
`;
